### ISSUE TITLE
Discussion: docker packaging, upgrading to latest dependencies

### ISSUE THREAD

**Issue by nyurik** [kartotherian/kartotherian#105]
Tilerator (and Kartotherian) hardcodes a number of packages that shouldn't really be part of it, e.g. `@kartotherian/cassandra`, `@kartotherian/autogen`, `@kartotherian/babel` and many other.  These packages are important for specific configurations, but not in the general case.  Additionally, Tilerator and other deps still use an old version of Mapnik ([current](https://www.npmjs.com/package/mapnik) is 4.21, we use 3.7).

I would like to propose a different packaging that will be both more flexible and leaner:
* `package.json` will not contain any optional packages
* specific targets, e.g. WMF build, would use additional `npm install <extra packages>` as part of the build process right after running `npm install` without any params.
* Docker-based build and deployment

This fits well into a more common modern packaging scenario - Docker/Kubernetes. Tilerator will be published as a "core tilerator" docker image in the `packages.docker.com`, plus the WMF-specific Dockerfile would use `npm install X` to add more components and publish it to the WMF-specific docker repository.

P.S. This will also allow us to have a more flexible, secure and smaller packages, e.g. by building Mapnik libraries during the build, but only copy the needed results to the resulting docker image.  Also, this method is still compatible with the original, because the resulting build results can still be pushed to the `deployment` git repo.

CC: @gehel @stephanebisson @mdholloway @mateusbs17

**Comment by thesocialdev** [comment]
Just a note, probably you already know: we can't have different node-mapnik versions across the packages. If some optional package has node-mapnik as a dependency it should be paired with the version kartotherian is using, otherwise, we can introduce breaking changes that will block the optional packages.

This could be a good opportunity to start mapping the dependencies architecture and maybe diminish the use of node-mapnik where it can be prevented and maybe concentrate the rendering functionalities in the core of Kartotherian/Tilerator.

[Here is an example](https://github.com/mapbox/makizushi/commit/8ffaa0415d78439d79ab3108ae36f845c789d475) of node-mapnik clean up in @mapbox/makizushi which needed only the blend functionality.

**Comment by nyurik** [comment]
Yes, mapnik version has to be the same in all. Where required, it should be a `peerDependency` rather than a regular `dependency`, with a wide range of allowed versions.  I don't think core should be aware of Mapnik at all -- Mapnik is (very slowly) loosing its grounds, and I can already make Tilerator generate tiles completely without it, and Kartotherian serve the vector tiles directly or rely on [mapbox-gl-native](https://github.com/mapbox/mapbox-gl-native/tree/master/platform/node) to render them.

**Comment by thesocialdev** [comment]
Mapnik is the maintenance burden. If we can replace it entirely, it would be excellent.

**Comment by nyurik** [comment]
@mateusbs17 it is, but we won't be able to get rid of all native components entirely.  The lib i mentioned is much smaller and much more recently developed and actively maintained, but it is still a native dependency.  The good thing is that it won't be needed on so many different levels of the system -- only in the vector -> raster rendering.
